### ISSUE TITLE
PixelFormat is not an enum, and ARV_PIXEL_FORMAT_* constants are not exposed

### ISSUE THREAD

**Issue by Ekleog** [robohouse-delft/aravis-rs#2]
I've just noticed that there is currently no clean way of defining pixel formats using only the rs library, as neither is `PixelFormat` an enum, nor are the `ARV_PIXEL_FORMAT_*` constants exposed. Right now, the best that is doable appears to be hand-copy-pasting the constant values from the aravis doc.

Anyway, as usual, thank you for this crate!

**Comment by de-vri-es** [comment]
The pixel formats are exposed in the mod `pixel_format`: https://docs.rs/aravis/0.0.3/aravis/pixel_formats/index.html

It can be a bit hard to find because of the structure generated by the gir tool. These are in the non-generated part of the crate. It is pretty much copy pasted, but atleast you don't have to copy paste it again :)

**Comment by EmmanuelP** [comment]
Hi Maarten,

Is there something we can do in Aravis that would help auto generate these contants ?

**Comment by de-vri-es** [comment]
Hey Emmanuel. Thanks for the offer. I really appreciate your responsiveness and friendly attitude :)

I don't think it can easily be solved by Aravis, because it's not really Aravis' fault. There are two reasons why I manually exposed the constants rather than letting the `gtk-rs/gir` tool do it:

1) The gir tool wants me to list all of the constants explicitly to export them from the safe crate. (Or maybe I couldn't even list them. I don't remember exactly which it was.)
2) I wanted to drop the `ARV_PIXEL_FORMAT_` prefix and instead use the `aravis::pixel_formats` module to namespace them. A `PIXEL_FORMAT_` prefix would be acceptable, but I'd still have to rename all constants manually in the configuration for the GIR tool.

Both points are a usability issue in the gir tool that Aravis can't fix. The second point is also related to rust concepts that don't really map to glib concepts (I think).

If the list of constants changes a lot, I would probably add a small script to this repo to auto-generate those in the current format. I'm pretty sure all required information is in the `.gir` file already :)

**Comment by EmmanuelP** [comment]
Ok. Thanks for the explanation.

Cheers.

**Comment by Ekleog** [comment]
Hmm… If `ArvPixelFormat` was an enumeration instead of a `guint32_t`, maybe that would solve the issue? On the other hand that's a breaking change I guess, but maybe it could still be done on the 0.7 branch?

**Comment by de-vri-es** [comment]
Turning it into an enum would also imply that it's a closed set. However, I think cameras are allowed to use custom formats. Especially in Rust, that would cause even more issues :p

**Comment by Ekleog** [comment]
If cameras can use custom formats, then it's quite a big issue indeed… in Rust it'd be possible to do it with `enum PixelFormat { Foo, Bar, Other(u32) }` I guess, but it's not really possible in C AFAIK.

**Comment by de-vri-es** [comment]
Well, C doesn't really enforce enum values to be one of the listed variants. They're basically just integers with a bunch of constants. It might be frowned upon to use enums that way though.

However, turning it into an enum still won't help with code generation for Rust. The gir tool won't automatically add an `Other(u32)` variant.

The binary representation of an enum with a `Other(u32)` variant will always be larger than a `u32`, since `Other` can overlap with the predefined constants. So `pixel_format as u32` won't work, and we would still need code to extract the `u32` from the enum. 

It may still be possible to manually create a `PixelFormat` struct with associated constants, instead of a separate module. It could also be an opaque type then, rather than an alias for `u32`. That could be a nice improvement for the safe API.

I'm thinking something along the lines of:
```rust
[repr(transparent)]
pub struct PixelFormat {
  raw: u32,
}

impl PixelFormat {
  pub const RGBA_8_PACKED : PixelFormat = ...;
  ...

  pub const fn custom(u32) -> PixelFormat {
    ...
  }
}
```

That would also help users discover the available values directly from the generated documentation. I'll experiment some more with this soon.

My main worry is that the automatically generated function wrappers that use `PixelFormat` won't compile with this.

**Comment by Ekleog** [comment]
Hmm yes, this totally makes sense indeed! Way better idea than my “enum with other” idea :)

I'm not sure the auto-generated function wrappers wouldn't compile with it — I seemed to remember that it was possible to tell to `gir` that the translation of a type was manually-handled? And given what you suggest is `#[repr(transparent)]`, it'd be ABI-compatible, so it not being the exact same type shouldn't be an issue.

**Comment by de-vri-es** [comment]
> I seemed to remember that it was possible to tell to `gir` that the translation of a type was manually-handled?

I had the same idea, and I tried to see if it could work on the [newtype-pirxel-format](https://github.com/robohouse-delft/aravis-rs/tree/newtype-pixel-format) branch.

But currently, you can only tell `gir` that classes are manually implemented, not primitive types or aliases for primitive types. I briefly investigated if it would be easy to patch `gir`, but I got a bit lost in the codebase.

This would still be nice to do though.

**Comment by Ekleog** [comment]
Hmm OK, that's problematic indeed :/ Thank you for having investigated into this! I've just opened https://github.com/gtk-rs/gir/issues/927 ; let's see how things move :)

**Comment by de-vri-es** [comment]
Thanks to your issue I managed to finish the newtype implementation. Turns out there is an undocumented `conversion_type` config key that can be used to enable calls to conversion traits in the generated code.

The end result is this: https://docs.rs/aravis/0.3.0/aravis/struct.PixelFormat.html

Thanks for driving this forward! This is much nicer than what we had before :D

It's released as `0.3.0` since it's a backwards incompatible change.
